### PR TITLE
Initial port of C++ tour

### LINKED ISSUE #23: Port A Whirlwind Tour of C++  to website
Port [A Whirlwind Tour of C++](https://www.cis.upenn.edu/~cis5470/cpp.doc) from google doc to a resource on our website.

## Tasks

- [ ] We want it to reside at: [software-analysis-class.org/resources/cpp](https://software-analysis-class.org/resources/cpp).
- [ ] We want the document stylized similar to other resources; ex. [course-vm](https://software-analysis-class.org/resources/course-vm)
- [ ] All links on the website to the old doc must be replaced with the above new url.
- [ ] All links on canvas to the old link should be replaced with the new url.

### PR DESCRIPTION
**Pull request by joshmhanson**
Made a few small edits to the original material, listed below.

**Under "Multiple Inheritance and Virtualism"**
added this at the end to clarify why we would use virtual inheritance / what it does
```
This structure ensures that a class `D` that inherits from `B` and `C` would
only contain one copy of `A`'s member variables. Read
[Understanding Virtual Tables in C++](https://pabloariasal.github.io/2017/06/10/understanding-virtual-tables/)
to gain a deeper understanding of how this works under the hood.
```

**Added "Local type inference"**
```
The type of variables can often be deduced automatically within a local
scope using the `auto` keyword. This mainly just allows C++ code to be
more concise, and can also save time when refactoring.
``` 

**Under "For-each loops"**

Added this at the end to show a full syntax example of type inference + for each loop + reference semantics

```
It is common to use for-each loops in combination with type inference
when the type of each element is too verbose to write out explicitly.
In some cases you can also specify reference semantics so that you're
accessing the elements themselves rather than copies of each element.

For example, after executing the below lines:

std::string s = "hello";
for (auto& c : s) {
    c = 'a';
}

the string `s` will be set to `"aaaaa"`.
```

### CODE REVIEW

**Comment by joshmhanson** [comment]
@hXtreme  Do I need to update the hyperlinks to use the same footnote style that was used in course-vm.markdown?

**Comment by hXtreme** [comment]
Yes, please use the same style as the course vm doc; it makes it easier to manage links and find out what all needs to be updated when we make changes.

File: resources/cpp.markdown [hunk 1]
```diff
@@ -0,0 +1,661 @@
...
+from the base class. It can also include its own member variables and
+functions. An example can be found [here](http://th.cpp.sh/6sv7cq).
+The Rectangle class is derived from the base class, Shape.
+
+Recall that derived classes can access protected members in the base
+class.
+
+As shown in the above-linked example, inheritance is specified using
+access modifiers `public`, `protected`, or `private`:
+```cpp
```
Comment (hXtreme) [inline resources/cpp.markdown:233]: can you add blank lines around the code block.

**Comment by hXtreme** [comment]
Looks pretty good, will do a proper review later but I think only minor styling edits may be required.

**Comment by joshmhanson** [comment]
When I was going through the cpp.sh links, I got warnings like this:
```
Short URLs have been disabled due to spam. You can now create (longer) links to your code using the "Link to this code" feature.
During some time, your code may still be reachable at http://th.cpp.sh/5m2
```
So for now I changed the links to th.cpp.sh. But the above message seems to suggest that these links may stop working at any point. The website also seems to be not updated since 2015 and doesn't have an HTTPS cert.

Some options:
- switch to replit, example here: https://replit.com/@hanjiexi/Overloading-Class-Constructors#main.cpp
- switch to OnlineGDB, example here: https://onlinegdb.com/iDyO5pvie
- find another similar site
- directly write the code onto the page; they can copy it into their own repl.it if they are following along as we suggest

File: resources/cpp.markdown (hunk 1 above)
Comment (joshmhanson) [reply resources/cpp.markdown:233]: done

**Comment by hXtreme** [comment]
Let's switch to using onlinegdb

**Comment by hXtreme** [comment]
@joshmhanson thanks for finding the issue with the links not working, I've now added a automated workflow that tries to search for broken links; though only internal links for now, I couldn't find a good way to check external links without unintentionally crawling the whole web, perhaps some day.

[![No broken links][no-broken-links/badge]][no-broken-links/workflow]
https://github.com/cis547/cis547.github.io/blob/master/.github/workflows/check-links.yml

[no-broken-links/workflow]: https://github.com/cis547/cis547.github.io/actions/workflows/check-links.yml
[no-broken-links/badge]: https://github.com/cis547/cis547.github.io/actions/workflows/check-links.yml/badge.svg

**Pull request merged**